subgraphs/mux-protocol/src/mappings/gmxHandlers.ts

```diff
@@ -0,0 +1,262 @@
+      collateralUSDDelta
+        .times(exponentToBigDecimal(collateralToken.decimals))
+        .div(collateralToken.lastPriceUSD!)
+    );
+  }
+
+  let positionSide = PositionSide.SHORT;
+  if (isLong) {
+    positionSide = PositionSide.LONG;
+  }
+  if (eventType == EventType.CollateralIn) {
+    const existingPosition = getUserPosition(
+      account,
+      pool,
+      collateralTokenAddress,
+      indexTokenAddress,
+      positionSide
+    );
+    if (!existingPosition) {
+      createPositionMap(
+        positionKey,
+        account,
+        pool,
+        collateralTokenAddress,
+        indexTokenAddress,
+        positionSide
+      );
+    }
+  }
+
+  let positionBalance = BIGINT_ZERO;
+  let positionBalanceUSD = BIGDECIMAL_ZERO;
+  let positionCollateralBalance = BIGINT_ZERO;
+  let positionCollateralBalanceUSD = BIGDECIMAL_ZERO;
+  const vaultContract = Vault.bind(event.address);
+  const tryGetPosition = vaultContract.try_getPosition(
+    Address.fromBytes(account.id),
+    collateralTokenAddress,
+    indexTokenAddress,
+    isLong
+  );
+  if (!tryGetPosition.reverted) {
+    positionBalanceUSD = convertToDecimal(
+      tryGetPosition.value.getValue0(),
+      GMX_PRICE_DECIMALS
+    );
+    positionCollateralBalanceUSD = convertToDecimal(
+      tryGetPosition.value.getValue1(),
+      GMX_PRICE_DECIMALS
+    );
+
+    const indexToken = getOrCreateToken(event, indexTokenAddress);
+    if (indexToken.lastPriceUSD && indexToken.lastPriceUSD! > BIGDECIMAL_ZERO) {
+      positionBalance = bigDecimalToBigInt(
+        positionBalanceUSD
+          .times(exponentToBigDecimal(indexToken.decimals))
+          .div(indexToken.lastPriceUSD!)
+      );
+    }
+    const collateralToken = getOrCreateToken(event, collateralTokenAddress);
+    if (
+      collateralToken.lastPriceUSD &&
+      collateralToken.lastPriceUSD! > BIGDECIMAL_ZERO
+    ) {
+      positionCollateralBalance = bigDecimalToBigInt(
+        positionCollateralBalanceUSD
+          .times(exponentToBigDecimal(collateralToken.decimals))
+          .div(collateralToken.lastPriceUSD!)
+      );
+    }
+  }
+
+  handleUpdatePositionEvent(
+    event,
+    pool,
+    account,
+    collateralToken,
+    collateralAmountDelta,
+    collateralUSDDelta,
+    positionCollateralBalance,
+    positionCollateralBalanceUSD,
+    indexToken,
+    sizeUSDDelta,
+    positionBalance,
+    positionBalanceUSD,
+    null,
+    isLong,
+    eventType,
+    convertToDecimal(liqudateProfit, GMX_PRICE_DECIMALS)
+  );
+}
+
+export function handleClosePosition(event: ClosePosition): void {
+  const pool = getOrCreateLiquidityPool(
+    event,
+    event.address,
+    GMX_POOL_NAME,
+    GMX_POOL_SYMBOL
+  );
+  takeSnapshots(event, pool);
+  const realisedPnlUSD = convertToDecimal(
+    event.params.realisedPnl,
+    GMX_PRICE_DECIMALS
+  );
+  updatePositionRealisedPnlUSD(event.params.key, realisedPnlUSD);
+
+  if (event.params.realisedPnl < BIGINT_ZERO) {
```

Comment:
If position is closed with a positive pnl, don't we have to update pool volume too? Since funds are leaving the pool to pay the trader?

Reply:
For GMX, every closePosition action will emit both ClosePosition() and DecreasePosition() event, so most of pool volumes computation has already been covered in DecreasePosition() event handler. As there is not pnl in DecreasePosition() event parameters, here in ClosePosition() event handler we just needs to cover the extra volumes when a users has a loss and has to use some of collateral to cover the loss. Thanks!

Reply:
Oh interesting, that makes sense. Could you add a comment there to make that clear for potential future readers?

Reply:
Added at 1f273b865a233a5b1744c9852222f87ddcf095b0